### PR TITLE
feat: math expressions in input fields

### PR DESCRIPTION
**Pull request by lucasser**
Implements #199

### CODE REVIEW

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/autocrafting/preview/AutocraftingPreviewScreen.java [hunk 1]
```diff
@@ -867,14 +869,18 @@ protected void onAmountFieldChanged() {
             return;
         }
         disableStartButton();
-        getAndValidateAmount().ifPresentOrElse(amount -> {
+        final AmountOperations.ReturnValue<Double> data = checkValue();
+        if (data.isValid()) {
             confirmButton.setMessage(PENDING);
-            changedAmount = amount;
+            changedAmount = data.getValue().get();
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/autocrafting/preview/AutocraftingPreviewScreen.java]: This can be much cleaner. Avoid usage of get() with optionals. Use ifPresentOrElse.

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/autocrafting/preview/AutocraftingPreviewScreen.java [hunk 2]
```diff
@@ -867,14 +869,18 @@ protected void onAmountFieldChanged() {
...
             amountField.setTextColor(0xFFFFFF);
-        }, () -> {
+        } else {
             confirmButton.setMessage(START);
             amountField.setTextColor(0xFF5555);
-        });
+        }
+        if (data.getValue().isPresent()) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/autocrafting/preview/AutocraftingPreviewScreen.java]: ifPresent(), don't use get()

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java [hunk 3]
```diff
@@ -157,6 +158,7 @@ private void addAmountField() {
         } else if (configuration.getInitialAmount() != null) {
             updateAmount(configuration.getInitialAmount());
         }
+        onAmountFieldChanged();
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java]: Why?

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java [hunk 4]
```diff
@@ -177,14 +179,30 @@ protected void onAmountFieldChanged() {
...
         } else {
             tryConfirm();
         }
         amountField.setTextColor(valid ? 0xFFFFFF : 0xFF5555);
+        if (data.getValue().isPresent()) {
+            setToolTip(formatToolTip(data.getValue().get()), amountField);
+        }
+    }
+
+    protected String formatToolTip(final N value) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java]:
```suggestion
    protected String formatCalculationTooltip(final N value) {
```

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java [hunk 5]
```diff
@@ -224,21 +242,25 @@ private void changeAmount(final int delta) {
         if (amountField == null) {
             return;
         }
-        getAndValidateAmount().ifPresentOrElse(oldAmount -> {
+        final AmountOperations.ReturnValue<N> value = checkValue();
+        final N newAmount;
+        if (value.isValid() && value.getValue().isPresent()) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java]: Why did you make this code worse with `isPresent()` and `get` calls. it was perfectly fine with ifPresentOrElse before.

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java [hunk 6]
```diff
@@ -305,15 +327,14 @@ protected void reset() {
...
     }
 
     private void tryConfirmAndCloseToParent() {
-        getAndValidateAmount().ifPresent(value -> {
-            if (confirm(value)) {
-                tryCloseToParent();
-            }
-        });
+        final AmountOperations.ReturnValue<N> value = checkValue();
+        if (value.isValid() && value.getValue().isPresent() && confirm(value.getValue().get())) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AbstractAmountScreen.java]: What is the use of `isValid` if we aready check if there is a value?

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java [hunk 7]
```diff
@@ -6,9 +6,30 @@
 public interface AmountOperations<N extends Number> {
     String format(N value);
 
-    Optional<N> parse(String value);
-
-    Optional<N> validate(N amount, @Nullable N minAmount, @Nullable N maxAmount);
+    ReturnValue<N> calculate(String value);
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]:
```suggestion
    ReturnValue<N> parse(String value);
```

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java [hunk 8]
```diff
@@ -6,9 +6,30 @@
...
 
     N changeAmount(@Nullable N current, int delta, @Nullable N minAmount, @Nullable N maxAmount);
+
+    class ReturnValue<N extends Number> {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]:
```suggestion
    class ParsedValue<N extends Number> {
```

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java [hunk 9]
```diff
@@ -6,9 +6,30 @@
...
+        private final @Nullable N value;
+        private final boolean valid;
+
+        public ReturnValue(@Nullable final N value, final boolean valid) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]: Why is value nullability and validity disjointed? I would expect that valid can be calculated if `value == null`

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/DoubleAmountOperations.java [hunk 10]
```diff
@@ -31,23 +30,14 @@ public String format(final Double value) {
...
 
     @Override
-    public Optional<Double> parse(final String value) {
+    public ReturnValue<Double> calculate(final String value) {
         try {
-            return Optional.of(Double.parseDouble(value));
+            return new ReturnValue<>(Double.parseDouble(value), true);
         } catch (final NumberFormatException e) {
-            return Optional.empty();
+            return new ReturnValue<>(false);
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/DoubleAmountOperations.java]:
```suggestion
            return ReturnValue.INVALID;
```

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/ExpressionAmountOperations.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    public String format(final Double value) {
+        return DECIMAL_FORMATTER.format(value);
+    }
+
+    @Override
+    public ReturnValue<Double> calculate(final String value) {
+        try {
+            final Evaluator calculator = new Evaluator(new Source(SOURCE, value));
+            return new ReturnValue<>(calculator.evaluate(), true);
+        } catch (EvaluatorException e) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/ExpressionAmountOperations.java]:
```suggestion
        } catch (final EvaluatorException e) {
```

**Review by raoulvdberge — CHANGES_REQUESTED** [review]

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java (hunk 9 above)
Comment (lucasser) [reply refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]: Value can also be invalid if it's out of bounds of the range, which is why I have to check for that instead of using ifpresentorelae in some places

**Comment by lucasser** [comment]
I added an ifValidOrElse method to ParsedValue to use instead of ifpresentorelse, idk if that works. Other than everything should be fixed.

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java [hunk 12]
```diff
@@ -1,14 +1,58 @@
...
-    Optional<N> validate(N amount, @Nullable N minAmount, @Nullable N maxAmount);
+    ParsedValue<N> parse(String value);
 
     N changeAmount(@Nullable N current, int delta, @Nullable N minAmount, @Nullable N maxAmount);
+
+    class ParsedValue<N extends Number> {
+        private final @Nullable N value;
+        private final boolean valid;
+
+        public ParsedValue(@Nullable final N value, final boolean valid) {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]: I don't like this constructor

So valid states are:
```
value == null && valid == false // invalid stuff like letters
value != null && valid == true   // normal state
value != null && valid == false  // errored state, there was a calculation but out of bounds possibly
```

That leaves
```
value == null && valid == true
```
Which is handled in this code, but really transparently.

Can you find a better more readable way to map these states?
I also don't like the `ifValid` and `ifValidOrElse` constructs.

**Review by raoulvdberge — CHANGES_REQUESTED** [review]

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java (hunk 12 above)
Comment (lucasser) [reply refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]: What is the problem with ifValid and ifValidOrElse?

Comment (raoulvdberge) [reply refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]: The way how you represent these states, and the states that are possible, are not apparent from the code.

Comment (lucasser) [reply refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java]: does this work better?

**Comment by raoulvdberge** [comment]
Please rebase on `develop` so I can get another look at this

File: refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java [hunk 13]
```diff
@@ -1,15 +1,54 @@
...
+        public ParsedValue(@Nullable final N value) {
+            this.value = value;
+            this.outOfBounds = false;
+        }
+
+        public Optional<N> getValue() {
+            return Optional.ofNullable(value);
+        }
+
+        public boolean isValid() {
```
Comment (raoulvdberge) [inline refinedstorage-common/src/main/java/com/refinedmods/refinedstorage/common/support/amount/AmountOperations.java:28]: Is the only use for `ParseValue` to indicate out of boundness? I don't care about reporting that error back to the user, so can we remove this class please?

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+/*
+* Evaluate a math expression inside a string.
+ */
+
+public class Evaluator {
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:29]: `MathExpressionEvaluator` and then we can remove the comment on top of this class.

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        "m", 1e6,
+        "g", 1e9,
+        "t", 1e12,
+        "p", 1e15,
+        "e", 1e18
+    );
+
+    public Evaluator(final Source source) {
+        if (source.content().isEmpty()) {
+            throw new EvaluatorException("");
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:43]: No message?

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        this.source = source;
+        final Lexer lexer;
+        try {
+            lexer = new Lexer(source, LexerTokenMappings.ARITHMETIC_MAPPINGS);
+            lexer.scan();
+        } catch (LexerException e) {
+            throw new EvaluatorException("unrecognized character");
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:51]: Always pass exception context along

```suggestion
            throw new EvaluatorException("unrecognized character", e);
```

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        final Parser parser;
+        try {
+            parser = new Parser(applyUnits(lexer.getTokens()), ParserOperatorMappings.ARITHMETIC_MAPPINGS);
+            parser.parse();
+        } catch (ParserException e) {
+            throw new EvaluatorException("syntax error");
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:58]: Context

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+                final TokenPosition position = new TokenPosition(source, new TokenRange(0, 0, 0, 0));
+                output.add(new Token("*", TokenType.BIN_OP, position));
+                try {
+                    final Double multiplier = unitMap.get(token.content().toLowerCase());
+                    final String stringMultiplier = String.valueOf(requireNonNull(multiplier));
+                    output.add(new Token(stringMultiplier, TokenType.FLOATING_NUMBER, position));
+                } catch (NullPointerException e) {
+                    throw new EvaluatorException("unrecognized unit " + token.content());
+                }
+                continue;
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:83]: Clean code: move to another method and add an early return if not nee to be transformed

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+    private double compute(final Node operand) {
+        switch (operand) {
+            case BinOpNode(Node left, Node right, Token binOp) -> {
+                return switch (binOp.content()) {
+                    case "+" -> compute(left) + compute(right);
+                    case "-" -> compute(left) - compute(right);
+                    case "*" -> compute(left) * compute(right);
+                    case "/" -> compute(left) / compute(right);
+                    case "%" -> compute(left) % compute(right);
+                    case "^" -> (Math.pow(compute(left), compute(right)));
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:103]: Useless `(` `)`

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+                }
+                return Double.parseDouble(token.content());
+            }
+            case ParenNode(List<Node> binOpNode) -> {
+                return compute(binOpNode.getFirst());
+            }
+            default -> {
+            }
+        }
+        throw new EvaluatorException(String.format("Unhandled node %s", operand));
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:119]: Shouldn't this be part of the default cause?

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/package-info.java [hunk 21]
```diff
@@ -0,0 +1,7 @@
+@ParametersAreNonnullByDefault
+@FieldsAndMethodsAreNonnullByDefault
+package com.refinedmods.refinedstorage.query.evaluator;
+
+import com.refinedmods.refinedstorage.api.core.FieldsAndMethodsAreNonnullByDefault;
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/package-info.java:5]: I don't think this exists in develop anymore?

File: CHANGELOG.md [hunk 22]
```diff
@@ -12,6 +12,7 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 ### Added
 
 -   You can now configure Refined Storage to not require energy. Controllers are not necessary in this case. Creative variants of items will be hidden and items will function without energy.
+-   Support for entering math expressions when requesting a craft, creating a pattern, and configuring a detector.
```
Comment (raoulvdberge) [inline CHANGELOG.md:15]:
```suggestion
-   Support for entering math expressions when requesting autocrafting, creating a processing pattern, and configuring a detector.
```

File: refinedstorage-query-parser/src/test/java/com/refinedmods/refinedstorage/query/evaluator/EvaluatorTest.java [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.refinedmods.refinedstorage.query.lexer.Source;
+
+import java.text.DecimalFormat;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class EvaluatorTest {
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/test/java/com/refinedmods/refinedstorage/query/evaluator/EvaluatorTest.java:11]: Where are the tests for the units?

File: refinedstorage-query-parser/src/test/java/com/refinedmods/refinedstorage/query/evaluator/EvaluatorTest.java [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+
+    private String buildAndExecute(final String expression) {
+        try {
+            final Evaluator math = new Evaluator(new Source(SOURCE_NAME, expression));
+            return df.format(math.evaluate());
+        } catch (EvaluatorException e) {
+            return e.getMessage();
+        }
+    }
+
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/test/java/com/refinedmods/refinedstorage/query/evaluator/EvaluatorTest.java:24]: I don't see any exception edge cases being tested. You probably want to run mutation testing here to check what you're missing!

File: refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+import static java.util.Objects.requireNonNull;
+
+/*
+* Evaluate a math expression inside a string.
+ */
+
+public class Evaluator {
+    private final Source source;
+    private final List<Node> nodes = new ArrayList<>();
+    private final Map<String, Double> unitMap = Map.of(
```
Comment (raoulvdberge) [inline refinedstorage-query-parser/src/main/java/com/refinedmods/refinedstorage/query/evaluator/Evaluator.java:32]: Can be static?

**Review by raoulvdberge — CHANGES_REQUESTED** [review]
And rebase on develop please

**Comment by raoulvdberge** [comment]
@lucasser Since I see PRs as a todo list for me to review, I'll close this PR for now. The review comments are there. Feel free to reopen if you have time to work on this again.
